### ISSUE TITLE
test_shutdown is not called when test_startup calls $self->test_skip

### ISSUE THREAD

**Issue by autarch** [test-class-moose/test-class-moose#59]
I'm pretty sure that this is a bug. Here's an example of why ...

At $WORK, we wrap the `test_startup` and `test_shutdown` methods to create and clear fixture data. This method wrapper kicks in *before* the `test_startup` in the class. This is necessary in case the individual class's `test_startup` needs the fixture data to exist.

There is a similar issue with `test_setup` and `test_teardown`, where a skipped method would not call `test_teardown`.

I think this should  be fixed, but it probably needs a big warning in Changes, since this is backwards-incompatible change and could conceivably break someone's test suite.

@Ovid, I'd appreciate your input on this.

**Comment by Ovid** [comment]
A couple of things here.

First, I've considered method modifiers which mutate state to be a serious code smell. For clients, I teach devs a standard pattern:

```
sub test_startup {
    my $test = shift;
    $test->next::method;
    ... # my setup here
}

sub test_setup {
    my $test = shift;
    $test->next::method;
    ... # my setup here
}

sub test_teardown {
    my $test = shift;
    ... # my setup here
    $test->next::method;
}

sub test_shutdown {
    my $test = shift;
    ... # my setup here
    $test->next::method;
}
```

With the above, we have parent test control methods called at the correct time.

Second, I tend to create fixtures by including roles and having the fixtures be lazily loaded inside of each test method. Thus, I don't have much to tear down. Though now that I stop to think about it, I use my `test_teardown` to do savepoint releases and rollbacks, so maybe I have skips somewhere which are allowing data to leak from one test to another ...

I should think that we don't want this change in behavior to be automatic. Make it a class or instance method to be called in the test class. Something like `ignore_skip_on_teardown()` or something like that, returning a true value. If you want that behavior, drop that into your base class. Or you could do it on a class by class basis, depending on your needs.

However, that's all just off the top of my head. I'll have to think more. Your thoughts, @autarch?

**Comment by autarch** [comment]
I'm not sure why you'd be using `next::method` instead of Moose method modifiers. Is there some reason to think that the method in the class is in a better place to decide whether a modifier should be called than vice versa (modifier can decide not to call the method)?

In our particular case, classes consume a "set up my fixtures" role that works by modifying all the control methods and doing the "right" things before or after the control method in question.

With what you're doing, we'd have to add the same (nearly empty) control methods to literally hundreds of test classes! Or maybe I'm misunderstanding.

Regardless, I think the idea of adding a method that controls whether the corresponding finalizer control method is called in the case of skips would make sense.

I'd probably call it `always_call_final_control_methods` and have it affect both `test_teardown` and `test_shutdown`. That seems a little simpler, and given that you're _either_ skipping a method or a class (never both together), it kind of just works.

**Comment by Ovid** [comment]
> I'm not sure why you'd be using next::method instead of Moose method modifiers.

Because method modifiers which modify the state of the class can leave it untrustworthy. Imagine, for example, a role with a modifier which adds 2 to a result. And another role with a modifier which divides a result by 5. If the original result was 3 then we have: `(3 + 2) / 5 == 1`. However, another programmer comes along and sorts the roles alphabetically and now you have `3 / 5 + 2 == 2.6`. In theory, the order in which we consume roles should not matter. When we add method modifiers which mutate state, the order can definitely matter.

That's why I use `$test->next::method`. It's explicit about the order in which things are called and I don't have mysterious action at a distance changing anything.

> With what you're doing, we'd have to add the same (nearly empty) control methods to literally hundreds of test classes! Or maybe I'm misunderstanding.

I just didn't explain well. Here's some actual production code:

```
package TestsFor::Veure::PartialClass::Character::Combat {
    use Test::Class::Moose extends => 'TestsFor::Veure';
    use namespace::autoclean;

    # need to fix these role names
    with qw(
      Test::Role::Fixture::Characters
      Test::Role::Fixture::StationArea
      Test::Role::Fixture::Station
      Test::Role::Fixture::Ships
    );

    sub test_attacking {
        my $test = shift;

        # these are lazily loaded via Test::Role::Fixture::Characters
        # so they won't be loaded if this method isn't called
        my $winston    = $test->character_winston;
        my $herve      = $test->character_npc_herve;
        my $talleyrand = $test->character_npc_talleyrand;
        ....
    }
}
```

In this case, we don't need `test_setup|test_teardown` for fixtures. Each test asks for the fixtures it needs.

**Comment by autarch** [comment]
> I just didn't explain well. Here's some actual production code:

Ah, I see. That's *very* different from how our code works, which looks more like this:

```
package TestFor::MM::Model::Thing;

use MM::Test::Class::Moose;

# use more modules ...

with 'TestRole::MM::Fixtures::HasMaxMind' => { populate => [ 'User', 'Payment', 'FoosballScore' ] };

**Comment by autarch** [comment]
So I was working on this today and realized that `test_teardown` is already called unconditionally, regardless of whether or not a method is skipped. The `test_shutdown` method is *not* called if the entire class is skipped.

@Ovid, is the fact that `test_teardown` is always called a bug? If not, this should be documented as the intended behavior, at least.

**Comment by autarch** [comment]
This is implemented in baf8f5db81c319c3205b0e2ae8f4fbe0e8e686f8 for the Test2 branch. I didn't think this was worth doing in the current master since testing it would be more work than I wanted to do ;)

**Comment by Ovid** [comment]
@autarch: sorry I missed the above query about `test_teardown`. I think that's a bug. There should probably be documented behavior and tests to back 'em up on that.

**Comment by autarch** [comment]
Ok, I can fix it so it's not always called. I'll do that in the Test2 branch, since it's so much easier to test these things there.

**Comment by autarch** [comment]
Resolved by baf8f5db81c319c3205b0e2ae8f4fbe0e8e686f8
